src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -381,4 +399,269 @@ public static Map<String, Map<String, WifiScanEntry>> getAggregatedWifiScans(
+	 * @param aggregatedFields	Which fields to be aggregated. RSSI, MCS for example.
+	 * @return Map<String, Map<String, Map<String, Map<JsonNode, Map<String, MeanAggregator>>>>>
+	 * 							Aggregated map that the keys are serial number, bssid, station
+	 * 							({@link State.Interface.SSID.Association#station}), aggregatedKey,
+	 * 							aggregatedField respectively. The returned value of type {@code
+	 * 							MeanAggregator} contains all the aggregation results in specified
+	 * 							fields.
+	 * @throws SecurityException
+	 * @throws NoSuchFieldException
+	 * @throws IllegalAccessException
+	 * @throws IllegalArgumentException
+	 */
+	public static Map<String, Map<String, Map<String, Map<JsonNode, Map<String, MeanAggregator>>>>> getAggregatedStates(
+		Modeler.DataModel dataModel,
+		long obsoletionPeriodMs,
+		long refTimeMs,
+		List<String> aggregatedKeys,
+		List<String> aggregatedFields
+	) throws IllegalArgumentException, IllegalAccessException,
+		NoSuchFieldException, SecurityException {
+		if (obsoletionPeriodMs < 0) {
+			throw new IllegalArgumentException(
+				"obsoletionPeriodMs must be non-negative."
+			);
+		}
+		Map<String, Map<String, Map<String, Map<JsonNode, Map<String, MeanAggregator>>>>> aggregatedStates =
+			new HashMap<>();
+
+		for (
+			Map.Entry<String, List<State>> deviceToStateList : dataModel.latestStates
+				.entrySet()
+		) {
+			String serialNumber = deviceToStateList.getKey();
+			List<State> states = deviceToStateList.getValue();
+
+			if (states.isEmpty()) {
+				continue;
+			}
+
+			Map<String, Map<String, Map<JsonNode, Map<String, MeanAggregator>>>> bssidToStats =
+				aggregatedStates
+					.computeIfAbsent(serialNumber, k -> new HashMap<>());
+
+			/**
+			 * Sort in reverse chronological order. Sorting is done just in case the
+			 * States in the original list are not chronological already - although
+			 * they are inserted chronologically, perhaps latency, synchronization, etc.
+			 */
+			states.sort(
+				(state1, state2) -> -Long.compare(state1.unit.localtime, state2.unit.localtime)
+			);
+			for (State state : states) {
+				if (refTimeMs - state.unit.localtime > obsoletionPeriodMs) {
+					// discard obsolete entries
+					continue;
+				}
+
+				for (Interface stateInterface : state.interfaces) {
+					if (stateInterface.ssids == null) {
+						continue;
+					}
+
+					for (SSID ssid : stateInterface.ssids) {
+						/**
+						 * clientsToStats is the mapping from the station to aggregated statistis.
+						 */
+						Map<String, Map<JsonNode, Map<String, MeanAggregator>>> clientToStats =
+							bssidToStats.computeIfAbsent(
+								ssid.bssid,
+								k -> new HashMap<>()
+							);
+
+						/**
+						 * aggregatedKey is used to Determine if two associations should be aggregated.
+						 * It is a JsonNode type with a mapping between channel/channel width/txpower
+						 * and the related values. The following condition is true then two stats results
```

Comment:
I think it is more clear to say "If two radios match in channel, channel widths, and txpower, then the two stats results need to be aggregated."